src/ucp/dt/datatype_iter.inl

```diff
@@ -138,32 +135,67 @@ ucp_datatype_iter_init(ucp_context_h context, void *buffer, size_t count,
                        int is_pack, ucp_datatype_iter_t *dt_iter,
                        uint8_t *sg_count, const ucp_request_param_t *param)
 {
-    size_t iov_length;
-    ucs_status_t status;
+    size_t length;
 
-    dt_iter->offset   = 0;
     dt_iter->dt_class = ucp_datatype_class(datatype);
+    dt_iter->offset   = 0;
 
     if (ucs_likely(dt_iter->dt_class == UCP_DATATYPE_CONTIG)) {
-        ucp_datatype_contig_iter_init(context, buffer, contig_length, dt_iter,
-                                      param);
         *sg_count = 1;
+        return ucp_datatype_contig_iter_init(context, buffer, contig_length,
+                                             dt_iter, param);
     } else if (dt_iter->dt_class == UCP_DATATYPE_IOV) {
-        iov_length = ucp_dt_iov_length((const ucp_dt_iov_t*)buffer, count);
-        status     = ucp_datatype_iov_iter_init(context, buffer, count,
-                                                iov_length, dt_iter, sg_count,
-                                                param);
-        if (status != UCS_OK) {
-            return status;
-        }
-    } else {
-        ucs_assert(dt_iter->dt_class == UCP_DATATYPE_GENERIC);
+        ucp_datatype_iter_iov_set_sg_count(sg_count, count);
+        length = ucp_dt_iov_length((const ucp_dt_iov_t*)buffer, count);
+        return ucp_datatype_iov_iter_init(context, buffer, count, length,
+                                          dt_iter, param);
+    } else if (!ENABLE_PARAMS_CHECK ||
```

Comment:
Why do we check `ENABLE_PARAMS_CHECK` here?
E.g. `ENABLE_PARAMS_CHECK = 0`, and `dt_iter->dt_class = 1 // strided`, `ucp_datatype_generic_iter_init` will be called.

Reply:
if ENABLE_PARAMS_CHECK==0 - we don't check for invalid dt_class and assume the 'else' case is generic
if ENABLE_PARAMS_CHECK==1 - we fail with UCS_ERR_INVALID_PARAM if dt_class is invalid